module/zfs/bpobj.c

```diff
@@ -206,29 +206,67 @@ bpobj_is_empty(bpobj_t *bpo)
 	    (!bpo->bpo_havesubobj || bpo->bpo_phys->bpo_num_subobjs == 0));
 }
 
+/*
+ * A recursive iteration of the bpobjs would be nice here but we run the risk
+ * of overflowing function stack space.  Instead, find each subobj and add it
+ * to the head of our list so it can be scanned for subjobjs.  Like the old
+ * recursive implementation, the "deepest" subobjs will be freed first.
+ * When a subobj is found to have no additional subojs, free it.
+ */
+typedef struct bpobj_info {
+	bpobj_t *bpi_bpo;
+	struct bpobj_info *bpi_parent;
+	uint64_t bpi_object;
+	uint64_t bpi_index;
+	list_node_t bpi_node;
+} bpobj_info_t;
+
+static bpobj_info_t *
+bpi_alloc(bpobj_t *bpo, bpobj_info_t *parent, uint64_t obj, uint64_t index)
+{
+	bpobj_info_t *bpi = kmem_zalloc(sizeof (bpobj_info_t), KM_SLEEP);
+	bpi->bpi_bpo = bpo;
+	bpi->bpi_parent = parent;
+	bpi->bpi_object = obj;
+	bpi->bpi_index = index;
+	return (bpi);
+}
+
+/*
+ * Update bpobj and all of its parents with new space accounting.
+ */
+static void
+propagate_space_reduction(bpobj_info_t *initial_bpi, uint64_t freed,
+    uint64_t comp_freed, uint64_t uncomp_freed)
+{
+	bpobj_info_t *bpi = initial_bpi;
+
+	while (bpi != NULL) {
+		bpobj_t *p = bpi->bpi_bpo;
+
+		p->bpo_phys->bpo_bytes -= freed;
```

Comment:
Like this..?
```
+boolean_t
+dmu_buf_is_dirty(dmu_buf_t *db_fake)
+{
+       dmu_buf_impl_t *db = (dmu_buf_impl_t *)db_fake;
+
+       return (db->db_dirtycnt > 0);
+}
+
```

Reply:
That seems fine, but we could be more specific by passing the tx, and then doing something like what we do in the beginning of dmu_buf_will_dirty() to see if it's already dirty.  Something like:

```
	mutex_enter(&db->db_mtx);
	for (dbuf_dirty_record_t *dr = db->db_last_dirty;
	    dr != NULL && dr->dr_txg >= tx->tx_txg; dr = dr->dr_next) {
		if (dr->dr_txg == tx->tx_txg) {
			mutex_exit(&db->db_mtx);
			return (B_TRUE);
		}
	}
	mutex_exit(&db->db_mtx);
        return (B_FALSE);
```

Reply:
Had some failures after trying this suggestion.
```
Presumably bpo_dbuf is already dirty, since you aren't dirtying it here. To make this code more obvious, I wonder if we should just eat the cost of calling dmu_buf_will_dirty() again here. Or we could add a dmu_buf_is_dirty() function and do ASSERT(dmu_buf_is_dirty(bpo_dbuf))?
```
```
loading concrete vdev 4, metaslab 12 of 15 ...
loading concrete vdev 4, metaslab 13 of 15 ...
loading concrete vdev 4, metaslab 14 of 15 ... dmu_buf_is_dirty(p->bpo_dbuf, tx) ASSERT at ../../module/zfs/bpobj.c:248:propagate_space_reduction()/var/lib/buildbot/slaves/zfs/Ubuntu_17_04_x86_64_Coverage__TEST_/build/zfs/tests/zfs-tests/tests/functional/alloc_class/alloc_class_012_pos.ksh[66]: log_must[67]: log_pos: line 253: 27193: Abort
```

So I'm going to have to explicitly call dbuf_will_dirty() and then we won't need the assert.

Reply:
That output looks like it's coming from zdb, which shouldn't be dirtying anything (or expecting it to be dirty).  So you may need to look at why the bpobj code is thinking that it's going to be changing the on-disk state when running zdb.

Reply:
When I moved the call to propagate_space_reduction out of the blkptr loop I had neglected to put if (free) in front of it.  I'm going to fix that and try again asserting that bpo_dbuf is already dirty.

Reply:
It's good now with if (free) propagate...(), and asserting dmu_buf_is_dirty.